Installation issue: `root` fails to build on MacOS due to GLEW link error

### Steps to reproduce the issue

On MacOS:

```console
$ spack install root
```


### Error message

<details><summary>Error message</summary>

<pre>
  >> 24436    ld: Undefined symbols:
     24437      ___GLXEW_ARB_multisample, referenced from:
     24438          TGLFormat::InitAvailableSamples() in TGLFormat.cxx.o
  >> 24439    clang: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 24440    make[2]: *** [graf3d/gl/CMakeFiles/RGL.dir/build.make:1521: lib/libRGL.6.32.08.so] Error
              1
     24441    make[2]: Leaving directory '/private/var/folders/hs/x6h90m5n60l2c5t6ckjfh9bw0000gp/T/ajpM
              ac/spack-stage/spack-stage-root-6.32.08-nv2iv3z4gr4spxy73hbz74bmysdhtzzu/spack-build-nv2i
              v3z'
  >> 24442    make[1]: *** [CMakeFiles/Makefile2:42964: graf3d/gl/CMakeFiles/RGL.dir/all] Error 2
     24443    make[1]: *** Waiting for unfinished jobs....
     24444    [ 96%] Building CXX object roofit/roofitcore/CMakeFiles/RooFitCore.dir/src/BidirMMapPipe.
              cxx.o
</pre></details>


### Information on your system

* **Spack:** 0.24.0.dev0 (fe746bdebb1f08962d286c4da37e88204942a813)
* **Python:** 3.12.1
* **Platform:** darwin-sonoma-cannonlake

### Additional information

PR #47483 re-introduces a bug that I had previously patched (but not properly fixed) in #45632: a link-time undefined symbols error appears in GLEW. The issue seems to arise from the fact that only ROOT's builtin version of GLEW seems to be properly linkable against; using a version of GLEW installed by Spack leads to the error shown above.

Pinging @drbenmorgan @gartung @greenc-FNAL @marcmengel @vitodb @vvolkl as the maintainers.

Pinging @sethrj as the author of #47483.

Pinging @paulgessinger as interested party in building ROOT on MacOS. 

Pinging @AJPfleger who encountered this problem as he was building ROOT on his Mac.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

> ROOT builds are always voluminous 😄 I'm looking at the ROOT builtin and see a recent change that may be related: [root-project/root@244161d](https://github.com/root-project/root/commit/244161d235b2b29f43d2482c3fdcbdd3338c222f)

I tried to incorporate this commit by adding a version locally:

```python
version("6.35.01", commit="33fa196189add043df120e085d6e8cd31ad3f74b")
```

But this yields the same problem:

```
     61695    cd /private/var/folders/7m/stmxrx597njc6hvnttm6xsp00000gr/T/sswatman/spack-stage/spack-stage-root-6.35.01-xswhtyzlchovpolukclqqrshpcclmtwc/spack-build-xswhtyz/graf3d/gl && /Users/sswatman/spack/opt
              /spack/darwin-sonoma-m2/apple-clang-15.0.0/cmake-3.30.2-kimc44cirqunadulq25k7ahu4m32rwqi/bin/cmake -E cmake_link_script CMakeFiles/RGL.dir/link.txt --verbose=1
     61696    /Users/sswatman/spack/lib/spack/env/clang/clang++  -Wc++11-narrowing -Wsign-compare -Wsometimes-uninitialized -Wconditional-uninitialized -Wheader-guard -Warray-bounds -Wcomment -Wtautological-comp
              are -Wstrncat-size -Wloop-analysis -Wbool-conversion -m64 -pipe -W -Wall -Woverloaded-virtual -fsigned-char -fno-common -Qunused-arguments -pthread -stdlib=libc++ -O3 -DNDEBUG -arch arm64 -isysroot
               /Library/Developer/CommandLineTools/SDKs/MacOSX15.1.sdk -dynamiclib -Wl,-headerpad_max_install_names -m64 -Wl,-dead_strip_dylibs -compatibility_version 6.35.0 -current_version 6.33.99 -o ../../lib
              /libRGL.6.33.99.so -install_name @rpath/libRGL.6.35.so CMakeFiles/RGL.dir/src/TArcBall.cxx.o CMakeFiles/RGL.dir/src/TF2GL.cxx.o CMakeFiles/RGL.dir/src/TGL5D.cxx.o CMakeFiles/RGL.dir/src/TGL5DDataSe
              tEditor.cxx.o CMakeFiles/RGL.dir/src/TGL5DPainter.cxx.o CMakeFiles/RGL.dir/src/TGLAdapter.cxx.o CMakeFiles/RGL.dir/src/TGLAnnotation.cxx.o CMakeFiles/RGL.dir/src/TGLAutoRotator.cxx.o CMakeFiles/RGL
              .dir/src/TGLAxis.cxx.o CMakeFiles/RGL.dir/src/TGLAxisPainter.cxx.o CMakeFiles/RGL.dir/src/TGLBoundingBox.cxx.o CMakeFiles/RGL.dir/src/TGLBoxPainter.cxx.o CMakeFiles/RGL.dir/src/TGLCamera.cxx.o CMak
              eFiles/RGL.dir/src/TGLCameraGuide.cxx.o CMakeFiles/RGL.dir/src/TGLCameraOverlay.cxx.o CMakeFiles/RGL.dir/src/TGLClip.cxx.o CMakeFiles/RGL.dir/src/TGLClipSetEditor.cxx.o CMakeFiles/RGL.dir/src/TGLCo
              ntext.cxx.o CMakeFiles/RGL.dir/src/TGLContextPrivate.cxx.o CMakeFiles/RGL.dir/src/TGLCylinder.cxx.o CMakeFiles/RGL.dir/src/TGLEmbeddedViewer.cxx.o CMakeFiles/RGL.dir/src/TGLEventHandler.cxx.o CMake
              Files/RGL.dir/src/TGLFaceSet.cxx.o CMakeFiles/RGL.dir/src/TGLFBO.cxx.o CMakeFiles/RGL.dir/src/TGLFontManager.cxx.o CMakeFiles/RGL.dir/src/TGLFormat.cxx.o CMakeFiles/RGL.dir/src/TGLH2PolyPainter.cxx
              .o CMakeFiles/RGL.dir/src/TGLHistPainter.cxx.o CMakeFiles/RGL.dir/src/TGLIsoMesh.cxx.o CMakeFiles/RGL.dir/src/TGLLegoPainter.cxx.o CMakeFiles/RGL.dir/src/TGLLightSet.cxx.o CMakeFiles/RGL.dir/src/TG
              LLightSetEditor.cxx.o CMakeFiles/RGL.dir/src/TGLLockable.cxx.o CMakeFiles/RGL.dir/src/TGLLogicalShape.cxx.o CMakeFiles/RGL.dir/src/TGLManip.cxx.o CMakeFiles/RGL.dir/src/TGLManipSet.cxx.o CMakeFiles
              /RGL.dir/src/TGLMarchingCubes.cxx.o CMakeFiles/RGL.dir/src/TGLObject.cxx.o CMakeFiles/RGL.dir/src/TGLOrthoCamera.cxx.o CMakeFiles/RGL.dir/src/TGLOutput.cxx.o CMakeFiles/RGL.dir/src/TGLOverlayButton
              .cxx.o CMakeFiles/RGL.dir/src/TGLOverlay.cxx.o CMakeFiles/RGL.dir/src/TGLPadPainter.cxx.o CMakeFiles/RGL.dir/src/TGLPadUtils.cxx.o CMakeFiles/RGL.dir/src/TGLParametric.cxx.o CMakeFiles/RGL.dir/src/
              TGLParametricEquationGL.cxx.o CMakeFiles/RGL.dir/src/TGLPerspectiveCamera.cxx.o CMakeFiles/RGL.dir/src/TGLPhysicalShape.cxx.o CMakeFiles/RGL.dir/src/TGLPlot3D.cxx.o CMakeFiles/RGL.dir/src/TGLPlotBo
              x.cxx.o CMakeFiles/RGL.dir/src/TGLPlotCamera.cxx.o CMakeFiles/RGL.dir/src/TGLPlotPainter.cxx.o CMakeFiles/RGL.dir/src/TGLPolyLine.cxx.o CMakeFiles/RGL.dir/src/TGLPolyMarker.cxx.o CMakeFiles/RGL.dir
              /src/TGLPShapeObj.cxx.o CMakeFiles/RGL.dir/src/TGLPShapeObjEditor.cxx.o CMakeFiles/RGL.dir/src/TGLPShapeRef.cxx.o CMakeFiles/RGL.dir/src/TGLQuadric.cxx.o CMakeFiles/RGL.dir/src/TGLRnrCtx.cxx.o CMak
              eFiles/RGL.dir/src/TGLRotateManip.cxx.o CMakeFiles/RGL.dir/src/TGLSAFrame.cxx.o CMakeFiles/RGL.dir/src/TGLSAViewer.cxx.o CMakeFiles/RGL.dir/src/TGLScaleManip.cxx.o CMakeFiles/RGL.dir/src/TGLSceneBa
              se.cxx.o CMakeFiles/RGL.dir/src/TGLScene.cxx.o CMakeFiles/RGL.dir/src/TGLSceneInfo.cxx.o CMakeFiles/RGL.dir/src/TGLScenePad.cxx.o CMakeFiles/RGL.dir/src/TGLSelectBuffer.cxx.o CMakeFiles/RGL.dir/src
              /TGLSelectRecord.cxx.o CMakeFiles/RGL.dir/src/TGLSphere.cxx.o CMakeFiles/RGL.dir/src/TGLStopwatch.cxx.o CMakeFiles/RGL.dir/src/TGLSurfacePainter.cxx.o CMakeFiles/RGL.dir/src/TGLText.cxx.o CMakeFile
              s/RGL.dir/src/TGLTF3Painter.cxx.o CMakeFiles/RGL.dir/src/TGLTH3Composition.cxx.o CMakeFiles/RGL.dir/src/TGLTransManip.cxx.o CMakeFiles/RGL.dir/src/TGLUtil.cxx.o CMakeFiles/RGL.dir/src/TGLViewerBase
              .cxx.o CMakeFiles/RGL.dir/src/TGLViewer.cxx.o CMakeFiles/RGL.dir/src/TGLViewerEditor.cxx.o CMakeFiles/RGL.dir/src/TGLVoxelPainter.cxx.o CMakeFiles/RGL.dir/src/TGLWidget.cxx.o CMakeFiles/RGL.dir/src
              /TH2GL.cxx.o CMakeFiles/RGL.dir/src/TH3GL.cxx.o CMakeFiles/RGL.dir/src/TKDEAdapter.cxx.o CMakeFiles/RGL.dir/src/TKDEFGT.cxx.o CMakeFiles/RGL.dir/src/TPointSet3DGL.cxx.o CMakeFiles/RGL.dir/src/TGLSd
              fFontMaker.cxx.o CMakeFiles/RGL.dir/src/TX11GL.cxx.o CMakeFiles/G__RGL.dir/G__RGL.cxx.o -F/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks  -Wl,-rpath,/Users/sswatman/s
              pack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/ftgl-2.4.0-po3enn4d5fcdcvu6j3u55dc222b7rhm7/lib -Wl,-rpath,/Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/gl2ps-1.4.2-brfnefu
              63o24vy35xunn43pkzcqgyazw/lib -Wl,-rpath,/private/var/folders/7m/stmxrx597njc6hvnttm6xsp00000gr/T/sswatman/spack-stage/spack-stage-root-6.35.01-xswhtyzlchovpolukclqqrshpcclmtwc/spack-build-xswhtyz/
              lib -Wl,-rpath,/Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/glew-2.2.0-k7xxv2onaymi2vy44hdvgcmpo2xqppvd/lib -framework OpenGL -framework OpenGL /Users/sswatman/spack/opt/spac
              k/darwin-sonoma-m2/apple-clang-15.0.0/ftgl-2.4.0-po3enn4d5fcdcvu6j3u55dc222b7rhm7/lib/libftgl.dylib /Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/gl2ps-1.4.2-brfnefu63o24vy35x
              unn43pkzcqgyazw/lib/libgl2ps.dylib /Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/libsm-1.2.4-qok2pubonu3ukcdmt5smqedvek44embi/lib/libSM.dylib /Users/sswatman/spack/opt/spack/d
              arwin-sonoma-m2/apple-clang-15.0.0/libice-1.1.1-6qudnqfkbyotbpn43n7bjnjedjoiyikl/lib/libICE.dylib /Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/libx11-1.8.9-borrydku5pkx5fcgiw
              5iowk5gayw2mlw/lib/libX11.dylib /Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/libxext-1.3.6-5p5qcdloxmtfcniq5bdm7vr6b7lviesy/lib/libXext.dylib ../../lib/libGed.6.33.99.so ../.
              ./lib/libRCsg.6.33.99.so ../../lib/libASImage.6.33.99.so -framework OpenGL /Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/glew-2.2.0-k7xxv2onaymi2vy44hdvgcmpo2xqppvd/lib/libGLE
              W.2.2.0.dylib -framework OpenGL ../../lib/libGui.6.33.99.so ../../lib/libTreePlayer.6.33.99.so ../../lib/libTree.6.33.99.so ../../lib/libGraf3d.6.33.99.so ../../lib/libGpad.6.33.99.so ../../lib/lib
              Postscript.6.33.99.so ../../lib/libGraf.6.33.99.so ../../lib/libHist.6.33.99.so ../../lib/libMatrix.6.33.99.so ../../lib/libMathCore.6.33.99.so ../../lib/libImt.6.33.99.so ../../lib/libMultiProc.6.
              33.99.so ../../lib/libNet.6.33.99.so ../../lib/libRIO.6.33.99.so ../../lib/libThread.6.33.99.so ../../lib/libCore.6.33.99.so
     61697    ld: warning: duplicate -rpath '/Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/glew-2.2.0-k7xxv2onaymi2vy44hdvgcmpo2xqppvd/lib' ignored
     61698    ld: warning: duplicate -rpath '/Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/gl2ps-1.4.2-brfnefu63o24vy35xunn43pkzcqgyazw/lib' ignored
     61699    ld: warning: duplicate -rpath '/Users/sswatman/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/ftgl-2.4.0-po3enn4d5fcdcvu6j3u55dc222b7rhm7/lib' ignored
  >> 61700    Undefined symbols for architecture arm64:
     61701      "___GLXEW_ARB_multisample", referenced from:
     61702          TGLFormat::InitAvailableSamples() in TGLFormat.cxx.o
     61703      "_glXChooseVisual", referenced from:
     61704          TGLWidget::CreateWindow(TGWindow const*, TGLFormat const&, unsigned int, unsigned int, std::__1::pair<void*, void*>&) in TGLWidget.cxx.o
     61705          TX11GLManager::InitGLWindow(unsigned long) in TX11GL.cxx.o
     61706      "_glXCreateContext", referenced from:

     ...

     61723          TGLContext::ClearCurrent() in TGLContext.cxx.o
     61724          TX11GLManager::MakeCurrent(int) in TX11GL.cxx.o
     61725      "_glXSwapBuffers", referenced from:
     61726          TGLContext::SwapBuffers() in TGLContext.cxx.o
     61727          TX11GLManager::Flush(int) in TX11GL.cxx.o
     61728    ld: symbol(s) not found for architecture arm64
  >> 61729    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 61730    make[2]: *** [lib/libRGL.6.33.99.so] Error 1
  >> 61731    make[1]: *** [graf3d/gl/CMakeFiles/RGL.dir/all] Error 2
     61732    make[1]: *** Waiting for unfinished jobs....
     61733    1 warning generated.
     61734    [ 98%] Built target G__RooFit
  >> 61735    make: *** [all] Error 2

```

--- a/var/spack/repos/builtin/packages/root/package.py
+++ b/var/spack/repos/builtin/packages/root/package.py
@@ -12,6 +12,8 @@ from spack.operating_systems.mac_os import macos_version
 from spack.package import *
 from spack.util.environment import is_system_path
 
+_is_macos = sys.platform == "darwin"
+
 
 class Root(CMakePackage):
     """ROOT is a data analysis framework."""
@@ -156,7 +158,7 @@ class Root(CMakePackage):
         when="@6.32.0:6.32.02",
     )
 
-    if sys.platform == "darwin":
+    if _is_macos:
         # Resolve non-standard use of uint, _cf_
         # https://sft.its.cern.ch/jira/browse/ROOT-7886.
         patch("math_uint.patch", when="@6.06.02")
@@ -186,7 +188,7 @@ class Root(CMakePackage):
     # See README.md for specific notes about what ROOT configuration
     # options are or are not supported, and why.
 
-    variant("aqua", default=False, description="Enable Aqua interface")
+    variant("aqua", default=_is_macos, description="Enable native macOS (Cocoa) interface")
     variant("arrow", default=False, description="Enable Arrow interface")
     variant("cuda", when="@6.08.00:", default=False, description="Enable CUDA support")
     variant("cudnn", when="@6.20.02:", default=False, description="Enable cuDNN support")
@@ -288,7 +290,7 @@ class Root(CMakePackage):
     variant(
         "webgui", default=True, description="Enable web-based UI components of ROOT", when="+root7"
     )
-    variant("x", default=True, description="Enable set of graphical options")
+    variant("x", default=(not _is_macos), description="Enable set of graphical options")
     variant("xml", default=True, description="Enable XML parser interface")
     variant("xrootd", default=False, description="Build xrootd file server and its client")
 
@@ -429,7 +431,7 @@ class Root(CMakePackage):
     conflicts("target=ppc64le:", when="@:6.24")
 
     # Incompatible variants
-    if sys.platform == "darwin":
+    if _is_macos:
         conflicts("+opengl", when="~x ~aqua", msg="root+opengl requires X or Aqua")
         # https://github.com/root-project/root/issues/7160
         conflicts("+aqua", when="~opengl", msg="+aqua requires OpenGL to be enabled")
@@ -455,15 +457,15 @@ class Root(CMakePackage):
     conflicts("%clang@16:", when="@:6.26.07", msg="clang 16+ support was added in root 6.26.08")
 
     # See https://github.com/spack/spack/pull/44826
-    if sys.platform == "darwin" and macos_version() == Version("12"):
+    if _is_macos and macos_version() == Version("12"):
         conflicts("@:6.27", when="+python", msg="macOS 12 python support for 6.28: only")
 
     # See https://github.com/root-project/root/issues/11714
-    if sys.platform == "darwin" and macos_version() >= Version("13"):
+    if _is_macos and macos_version() >= Version("13"):
         conflicts("@:6.26.09", msg="macOS 13 support was added in root 6.26.10")
 
     # See https://github.com/root-project/root/issues/16219
-    if sys.platform == "darwin" and macos_version() >= Version("15"):
+    if _is_macos and macos_version() >= Version("15"):
         conflicts("@:6.32.05", msg="macOS 15 support was added in root 6.32.06")
 
     # ROOT <6.14 is incompatible with Python >=3.7, which is the minimum supported by spack
@@ -627,8 +629,6 @@ class Root(CMakePackage):
         # Options related to ROOT's ability to download and build its own
         # dependencies. Per Spack convention, this should generally be avoided.
 
-        afterimage_enabled = ("+x" in self.spec) if "platform=darwin" not in self.spec else True
-
         options += [
             define("builtin_cfitsio", False),
             define("builtin_davix", False),
@@ -655,7 +655,12 @@ class Root(CMakePackage):
         ]
 
         if self.spec.satisfies("@:6.32"):
-            options.append(define("builtin_afterimage", afterimage_enabled))
+            options.append(
+                define(
+                    "builtin_afterimage",
+                    ("+x" in self.spec) if "platform=darwin" not in self.spec else True,
+                )
+            )
 
         # Features
         options += [
@@ -764,7 +769,7 @@ class Root(CMakePackage):
 
         # #################### Compiler options ####################
 
-        if sys.platform == "darwin" and self.compiler.cc == "gcc":
+        if _is_macos and self.compiler.cc == "gcc":
             cflags = "-D__builtin_unreachable=__builtin_trap"
             options.extend([define("CMAKE_C_FLAGS", cflags), define("CMAKE_CXX_FLAGS", cflags)])
 

